File: scripts/bridge_js_source.py
"""Bridge JS source text for the static API contract checks.

js/src/llama_webgpu_bridge.js is only the public entry; the implementation
lives in the .js and .ts modules it imports (the public .d.ts is excluded). The
checks read every module, joined in the order the former single-file source
declared them (helpers, worker host and proxy, direct runtime, facade), so
multi-token patterns keep their scope. A missing listed module raises instead
of silently narrowing a check.
"""
from __future__ import annotations

from pathlib import Path

_ORDERED_MODULES = (
    "worker_protocol.ts",
    "worker_host.js",
    "worker_proxy.js",
    "runtime.js",
    "bridge.js",
    "llama_webgpu_bridge.js",
)


def _modules(directory: Path, pattern: str) -> list[Path]:
    return [
        path
        for suffix in (".js", ".ts")
        for path in directory.glob(pattern + suffix)
        if not path.name.endswith(".d.ts")
    ]


def bridge_js_source(root: Path) -> str:
    source_dir = root / "js" / "src"
    internal = sorted(_modules(source_dir / "internal", "*"))
    ordered = [source_dir / name for name in _ORDERED_MODULES]
    listed = {*internal, *ordered}
    remaining = sorted(
        path for path in _modules(source_dir, "**/*") if path not in listed
    )
    return "\n".join(
        path.read_text(encoding="utf-8")
        for path in (*internal, *ordered, *remaining)
    )


def method_body(source: str, signature: str) -> str:
    """Returns a class method from its signature through its closing brace.

    Scoping an ordered pattern to one method keeps it from matching text in a
    later method. A missing signature returns "", so the check fails closed.
    """
    start = source.find(f"\n  {signature}")
    if start < 0:
        return ""
    end = source.find("\n  }\n", start)
    return source[start : end + len("\n  }") if end > 0 else len(source)]

File: tests/bridge/bridge_js_source.mjs
// Bridge JS source text for the static API contract tests. js/src/llama_webgpu_bridge.js is only the public
// entry; the implementation lives in the .js and .ts modules it imports (the
// public .d.ts is excluded). The tests read every module, joined in the order
// the former single-file source declared them (helpers, worker host and proxy,
// direct runtime, facade), so multi-token patterns keep their scope. A missing
// listed module throws instead of silently narrowing a check.
import { readFileSync, readdirSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

export const repoRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');

const ORDERED_MODULES = [
  'worker_protocol.ts',
  'worker_host.ts',
  'worker_proxy.ts',
  'runtime.ts',
  'bridge.ts',
  'llama_webgpu_bridge.js',
];

// Reads a repository file as Python's read_text() does: universal newlines.
export function readRepoText(relativePath) {
  return readFileSync(path.join(repoRoot, relativePath), 'utf8').replace(/\r\n?/g, '\n');
}

function isModule(name) {
  return (name.endsWith('.js') || name.endsWith('.ts')) && !name.endsWith('.d.ts');
}

// Python sorts paths part by part, not as flat strings.
function comparePaths(left, right) {
  const a = left.split('/');
  const b = right.split('/');
  for (let index = 0; index < Math.min(a.length, b.length); index += 1) {
    if (a[index] !== b[index]) {
      return a[index] < b[index] ? -1 : 1;
    }
  }
  return a.length - b.length;
}

export function bridgeJsSource() {
  const sourceDir = 'js/src';
  const internal = readdirSync(path.join(repoRoot, sourceDir, 'internal'), { withFileTypes: true })
    .filter((entry) => entry.isFile() && isModule(entry.name))
    .map((entry) => `${sourceDir}/internal/${entry.name}`)
    .sort(comparePaths);
  const ordered = ORDERED_MODULES.map((name) => `${sourceDir}/${name}`);
  const listed = new Set([...internal, ...ordered]);
  const remaining = readdirSync(path.join(repoRoot, sourceDir), { recursive: true, withFileTypes: true })
    .filter((entry) => entry.isFile() && isModule(entry.name))
    .map((entry) => path.relative(repoRoot, path.join(entry.parentPath, entry.name)).split(path.sep).join('/'))
    .filter((relativePath) => !listed.has(relativePath))
    .sort(comparePaths);
  return [...internal, ...ordered, ...remaining].map(readRepoText).join('\n');
}

// Returns a class method from its signature through its closing brace.
// Scoping an ordered pattern to one method keeps it from matching text in a
// later method. A missing signature returns "", so the check fails closed.
export function methodBody(source, signature) {
  const start = source.indexOf(`\n  ${signature}`);
  if (start < 0) {
    return '';
  }
  const end = source.indexOf('\n  }\n', start);
  return source.slice(start, end > 0 ? end + '\n  }'.length : source.length);
}
